server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JMAPApiRoutes.scala

```
@@ -97,26 +98,34 @@ class JMAPApiRoutes (val authenticator: Authenticator,
     }
 
   private def process(requestObject: RequestObject,
                       httpServerResponse: HttpServerResponse,
-                      mailboxSession: MailboxSession): SMono[Void] =
-    requestObject
-      .methodCalls
-      .map(invocation => this.processMethodWithMatchName(requestObject.using.toSet, invocation, mailboxSession))
-      .foldLeft(SFlux.empty[Invocation]) { (flux: SFlux[Invocation], mono: SMono[Invocation]) => flux.mergeWith(mono) }
-      .collectSeq()
-      .flatMap((invocations: Seq[Invocation]) =>
-        SMono.fromPublisher(httpServerResponse.status(OK)
-          .header(CONTENT_TYPE, JSON_CONTENT_TYPE)
-          .sendString(
-            SMono.fromCallable(() =>
-              serializer.serialize(ResponseObject(ResponseObject.SESSION_STATE, invocations)).toString),
-            StandardCharsets.UTF_8
-          ).`then`())
-      )
+                      mailboxSession: MailboxSession): SMono[Void] = {
+    val unsupportedCapabilities = requestObject.using
+      .filterNot(capability => CapabilityIdentifier.ALL_CAPABILITIES.contains(capability))
+
+    if (unsupportedCapabilities.isEmpty) {
+      requestObject
+        .methodCalls
+        .map(invocation => this.processMethodWithMatchName(requestObject.using.toSet, invocation, mailboxSession))
+        .foldLeft(SFlux.empty[Invocation]) { (flux: SFlux[Invocation], mono: SMono[Invocation]) => flux.mergeWith(mono) }
+        .collectSeq()
+        .flatMap((invocations: Seq[Invocation]) =>
+          SMono.fromPublisher(httpServerResponse.status(OK)
+            .header(CONTENT_TYPE, JSON_CONTENT_TYPE)
+            .sendString(
+              SMono.fromCallable(() =>
+                serializer.serialize(ResponseObject(ResponseObject.SESSION_STATE, invocations)).toString),
+              StandardCharsets.UTF_8
+            ).`then`())
+        )
+    } else {
+      SMono.raiseError(UnsupportedCapabilitiesException(unsupportedCapabilities))
```

you could fail fast instead, it would improve readability of this long function